store/tikv/coprocessor.go

```diff
@@ -1189,3 +1225,151 @@ func (it copErrorResponse) Next(ctx context.Context) (kv.ResultSubset, error) {
 func (it copErrorResponse) Close() error {
 	return nil
 }
+
+type rateLimitAction struct {
+	enabled        bool
+	fallbackAction memory.ActionOnExceed
+	once           sync.Once
+	// totalTokenNum indicates the total token at initial
+	totalTokenNum uint
+	cond          struct {
+		*sync.Cond
+		// exceed indicates whether have encountered OOM situation.
+		exceed bool
+		// existedTokenNum indicates the count of tokens which still exists
+		existedTokenNum uint
+		// isTokenDestroyed indicates whether there is one token has been isTokenDestroyed after Action
+		isTokenDestroyed bool
+	}
+}
+
+func newRateLimitAction(totalTokenNumber uint, cond *sync.Cond) *rateLimitAction {
+	return &rateLimitAction{
+		once:          sync.Once{},
+		totalTokenNum: totalTokenNumber,
+		cond: struct {
+			*sync.Cond
+			exceed           bool
+			existedTokenNum  uint
+			isTokenDestroyed bool
+		}{
+			Cond:            cond,
+			exceed:          false,
+			existedTokenNum: totalTokenNumber,
+		},
+	}
+}
+
+// Action implements ActionOnExceed.Action
+func (e *rateLimitAction) Action(t *memory.Tracker) {
+	if !e.enabled {
+		if e.fallbackAction != nil {
+			e.fallbackAction.Action(t)
+		}
+		return
+	}
+
+	e.once.Do(func() {
+		e.conditionLock()
+		defer e.conditionUnlock()
+		if e.cond.existedTokenNum < 2 {
+			logutil.BgLogger().Debug("rateLimitAction delegate to fallback action",
+				zap.Uint("totalTokenNum", e.totalTokenNum))
+			if e.fallbackAction != nil {
+				e.fallbackAction.Action(t)
+			}
+			return
+		}
```

Comment:
If `e.cond.existedTokenNum < 2` , then the action will fallback only once because e.once will never reset?

Reply:
If `e.cond.existedTokenNum >= 2` and read data finished, can this action also fallbacks if OOM?

Reply:
> If e.cond.existedTokenNum >= 2 and read data finished, can this action also fallbacks if OOM?

`rateLimitAction.enabled` will handle this scenario

Reply:
If all data has been read, we should close `rateLimitAction` immediately

Reply:
This comment is not addressed @Yisaer